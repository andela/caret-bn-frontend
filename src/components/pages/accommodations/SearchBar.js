import React, { Component } from 'react';
import {
  Col, Button, Row, Form,
} from 'react-bootstrap';
import { connect } from 'react-redux';
import { showAlert } from '../../../actions/alertAction';
import AlertComponent from '../../global/AlertComponent';
import { getLocations } from '../../../actions/locationActions';
import { accommodationSearch } from '../../../actions/accommodationActions';

class SearchBar extends Component {
  state = {
    params: null,
    error: {
      status: false,
      heading: null,
      message: null,
    },
  }

  componentDidMount = async () => {
    const { getLocations } = this.props;
    await getLocations();
  }

  submitSearch = async () => {
    const { params } = this.state;
    const {
      showAlert, startSearch, accommodationSearch, stopLoader,
    } = this.props;
    await this.setState({
      error: {
        status: false,
        heading: null,
      },
    });
    if (params === null) {
      await this.setState({
        error: {
          status: true,
          heading: 'Empty Search',
          message: 'You need to enter details',
        },
      });
      return showAlert();
    }
    let searchParams = '';
    Object.keys(params).forEach((key) => {
      if (params[`${key}`] !== '') {
        searchParams += `${key}=${params[`${key}`]}&`;
      }
    });
    searchParams = searchParams.substring(0, searchParams.length - 1);
    if (searchParams !== '') {
      searchParams = `?${searchParams}`;
    }
    startSearch();
    await accommodationSearch(searchParams);
    stopLoader();
  };

  handleChange = async (e) => {
    const { name, value } = e.target;
    await this.setState((prevState) => ({
      params: {
        ...prevState.params,
        [name]: value,
      },
    }));
  }

  render() {
    const { locations, endSearch } = this.props;
    const { error } = this.state;
    return (
      <Row style={{ margin: '10px' }}>
        <Form>
          <Row className="center-items">
            {
              (error.status) ? <AlertComponent variant="danger" heading={error.heading} message={error.message} /> : ''
            }
          </Row>
          <Row>
            <Col xs={12} sm={12} md={6} lg={6}>
              <Form.Group>
                <Form.Control as="select" name="location" data-test="location" onChange={(e) => this.handleChange(e)}>
                  <option selected disabled>Select Location...</option>
                  <option value="">None</option>
                  {
                    locations ? locations.data.map((location) => <option key={location.id} value={location.name}>{location.name}</option>) : null
                  }
                </Form.Control>
              </Form.Group>
            </Col>
            <Col xs={12} sm={12} md={6} lg={6}>
              <Form.Group>
                <Form.Control as="input" name="name" data-test="name" placeholder="Name" onChange={(e) => this.handleChange(e)} />
              </Form.Group>
            </Col>
            <Col xs={12} sm={12} md={6} lg={6}>
              <Form.Group>
                <Form.Control as="input" name="description" data-test="description" placeholder="Description" onChange={(e) => this.handleChange(e)} />
              </Form.Group>
            </Col>
            <Col xs={12} sm={12} md={6} lg={6}>
              <Form.Group>
                <Form.Control as="input" name="highlights" data-test="highlights" placeholder="Highlights" onChange={(e) => this.handleChange(e)} />
              </Form.Group>
            </Col>

            <Col xs={12} sm={12} md={6} lg={6}>
              <Form.Group>
                <Form.Control as="input" name="amenities" data-test="amenities" placeholder="Amenities" onChange={(e) => this.handleChange(e)} />
              </Form.Group>
            </Col>
            <Col xs={12} sm={12} md={6} lg={6}>
              <Form.Group>
                <Form.Control as="input" name="rating" data-test="rating" placeholder="Rating" onChange={(e) => this.handleChange(e)} />
              </Form.Group>
            </Col>
          </Row>
          <Row className="center-items">
            <Button className="full-width-buttons" data-test="submit-button" onClick={() => this.submitSearch()}>Search</Button>
            <Button className="full-width-buttons" data-test="show-all-button" onClick={() => endSearch()}>Show All</Button>
          </Row>
        </Form>
      </Row>
    );
  }
}

const mapStateToProps = (state) => ({
  locations: state.locations.data,
});

export default connect(mapStateToProps, { getLocations, showAlert, accommodationSearch })(SearchBar);
