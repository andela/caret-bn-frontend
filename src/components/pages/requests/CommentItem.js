/* eslint-disable react/button-has-type */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import {
  Container, Row, Col, Button, Form,
} from 'react-bootstrap';
import {
  getComment, deleteComment, editComment,
} from '../../../actions/commentAction';
import { singleRequestAction } from '../../../actions/requestsActions';
import Confirm from '../../global/Confirm';
import trash from '../../../assets/images/trash.svg';
import edit from '../../../assets/images/edit.png';
import cancel from '../../../assets/images/cancel.jpg';
import isAuthenticated from '../../../helpers/isAuthenticated';

export class CommentItem extends Component {
    state = {
      isLoading: false,
      openComment: false,
      comment: '',
      isProcessing: false,
      editMode: false,
      commentId: null,
      draftComment: null,
      loadingButton: false,
      loadingButtonEdit: false,
      deleteLoading: false,
    };

    UNSAFE_componentWillReceiveProps(prevProps) {
      const { editData } = prevProps;
      if (editData === true) {
        this.setState({ draftComment: null, comment: '', editMode: false });
      }
    }

    handleChange = (e) => {
      const { name, value } = e.target;
      this.setState({
        [name]: value,
      });
    };

    actionSwitch = async (action, commentId) => {
      this.setState({
      });

      switch (action) {
        case 'delete':
          this.dispatchDeleteComment(commentId);
          break;
        default:
          this.dispatchRejectBooking(commentId);
      }
    }

    dispatchDeleteComment = async (commentId) => {
      const { requestId, deleteComment, getComment } = this.props;

      this.setState({
        operation: 'delete',
        deleteLoading: true,
      });
      await this.props.deleteComment(commentId);
      this.setState({
        deleteLoading: false,
      });
      await this.props.getComment(requestId);
      this.setState({
        isProcessing: false,
      });
    }

    editButton = (e) => {
      e.preventDefault();

      const { comment, id } = e.target;
      this.setState({ draftComment: comment, commentId: id }, () => {
        this.setState({ editMode: true });
      });
    }

    sendCommentEdit = async (e) => {
      e.preventDefault();
      const { requestId, getComment } = this.props;
      const { commentId, draftComment } = this.state;
      const comment = { comment: draftComment };

      this.setState({ loadingButtonEdit: true });
      await this.props.editComment(comment, commentId);
      await this.props.getComment(requestId);
      this.setState({ editMode: false, loadingButtonEdit: false });
    };

    closeEditMode = (e) => {
      e.preventDefault();
      this.setState({ draftComment: null, comment: '', editMode: false });
    };


    render() {
      const { element, userInfo } = this.props;

      const {
        isLoading, openComment, comment, commentId, editMode, draftComment, loadingButton, loadingButtonEdit, deleteLoading,
      } = this.state;

      return (
            <div data-test="commentcard" className="commentCard p-2" key={element.id}>
             <div className="d-flex mb-1">
             <img src={element.user.image} className="icon menu-photo" />
             <div className="d-flex flex-column ml-3">
              <p className="small py-0 mb-0 ">{element.user.username}</p>
              <p className="small py-0 mb-0 text-secondary"><small>{moment(element.updatedAt).fromNow()}</small></p>
             </div>
             </div>
             <div className="d-flex justify-content-between align-items-center">
              {editMode && commentId == element.id ? (
              <form className="w-100" onSubmit={this.sendCommentEdit}>
                <input type="text" className="form-control" name="draftComment" value={draftComment} onChange={this.handleChange} required />
                <Button type="submit">{loadingButtonEdit ? <i style={{ fontSize: '20px' }} className="fas fa-spinner fa-pulse" /> : 'save'}</Button>
              </form>
              ) : (
              <p className="font-weight-light w-100 m-0 p-0">{deleteLoading ? <i style={{ fontSize: '20px' }} className="fas fa-spinner fa-pulse" /> : element.comment}</p>
              )}
              <div className="commentTrash d-flex m-0 p-1">
              { element.user.email === userInfo.payload.email
                ? (
                  <>
                  {editMode === true && commentId == element.id ? (
                    <button className="btn">
                    <img
                      onClick={(e) => {
                        this.closeEditMode(e);
                      }}
                      src={cancel}
                      alt=""
                      className="bodyIcons trash"
                    />
                    </button>
                  ) : (
                    <button className="btn">
                    <img
                      onClick={(e) => {
                        e.target.comment = element.comment;
                        e.target.id = element.id;
                        this.editButton(e);
                      }}
                      src={edit}
                      alt=""
                      className="bodyIcons footerIconsRight edit"
                    />
                    </button>
                  )}
              <Confirm
                data-test="approve"
                variant="danger-outline"
                action="delete"
                id={element.id}
                processAction={this.actionSwitch}
                title={(
              <img
                src={trash}
                alt=""
                className="bodyIcons footerIconsRight trash"
              />
            )}
                size="sm"
                buttonClass="process-request-button btn-block"
              />
                  </>
                )
                : ''}
              </div>
             </div>
            </div>
      );
    }
}
export const mapStateToProps = (state) => ({
  dataError: state.requestComment.dataError,
  data: state.requestComment.data,
  commentError: state.requestComment.commentError,
  commentData: state.requestComment.commentData,
  deleteData: state.requestComment.deleteData,
  deleteError: state.requestComment.deleteError,
  editData: state.requestComment.editData,
  editError: state.requestComment.editError,
  status: state.requestComment.status,
  profile: state.profile.data.profile,
});

export default connect(mapStateToProps, {
  getComment, deleteComment, singleRequestAction, editComment,
})(CommentItem);
