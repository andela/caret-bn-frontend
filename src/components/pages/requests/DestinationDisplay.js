import React from 'react';
import { Row, Form, Col } from 'react-bootstrap';
import PropTypes from 'prop-types';

const DestinationDisplay = ({
  destination, index, disabled, location, departureDisable, onChange,
  id,
  arrivalDate,
  departureDate,
  reasons,
}) => (
  <span data-test="destination-display" className="destination-wrapper">
    <Col xs={12} sm={12} md={4} lg={4}>
      <Form.Group>
        <Form.Label>
            Destination
            {' '}
            {index + 1}
        </Form.Label>
        <Form.Control
          as="select"
          disabled
          onChange={onChange}
          name={`id-${index}`}
          id={`id-${index}`}
        >
          <option value={destination.location.id}>{destination.location.name}</option>
          {location}
        </Form.Control>
      </Form.Group>
    </Col>

    <Col xs={12} sm={12} md={4} lg={4}>
      <Form.Group>
        <Form.Label>Arrival Date</Form.Label>
        <Form.Control
          as="input"
          type="date"
          value={arrivalDate}
          disabled={disabled}
          onChange={onChange}
          name={`arrivalDate-${index}`}
          id={`arrivalDate-${index}`}
        />
      </Form.Group>
    </Col>

    <Col xs={12} sm={12} md={4} lg={4}>
      <Form.Group>
        <Form.Label>Departure Date</Form.Label>
        <Form.Control
          as="input"
          type="date"
          value={departureDate}
          disabled={departureDisable}
          onChange={onChange}
          name={`departureDate-${index}`}
          id={`departureDate-${index}`}
        />
      </Form.Group>
    </Col>

    <Col xs={12} sm={12} md={9} lg={9}>
      <Form.Group>
        <Form.Label>Reasons</Form.Label>
        <Form.Control
          as="textarea"
          value={reasons}
          disabled={disabled}
          onChange={onChange}
          name={`reasons-${index}`}
          id={`reasons-${index}`}
        />
      </Form.Group>
    </Col>
  </span>
);

DestinationDisplay.propTypes = {
  destination: PropTypes.object.isRequired,
  index: PropTypes.number,
};

export default DestinationDisplay;
