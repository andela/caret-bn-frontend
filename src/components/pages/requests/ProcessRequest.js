import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button } from 'react-bootstrap';
import PropTypes from 'prop-types';
import { processRequestAction, singleRequestAction } from '../../../actions/requestsActions';
import { getManagerRequestAction } from '../../../actions/managerRequestAction';

export class ProcessRequest extends Component {
  state = {
    isLoading: false,
  };

  processAction = async () => {
    const { props } = this;
    const { action, id } = props;

    this.setState({ isLoading: true });

    await props.processRequestAction(action, id);
    await props.getManagerRequestAction();
    await props.singleRequestAction(id);

    this.setState({ isLoading: false });
  };

  render() {
    const { isLoading } = this.state;
    const { action, variant, disabled } = this.props;

    return (
      <Button data-test="process-request-button" className="process-request-button btn-block" onClick={this.processAction} variant={variant} disabled={disabled}>
        { isLoading ? <i className="fas fa-spinner fa-pulse" /> : action }
      </Button>
    );
  }
}

ProcessRequest.propTypes = {
  action: PropTypes.string.isRequired,
  variant: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  id: PropTypes.number.isRequired,
  processRequestAction: PropTypes.func,
  getManagerRequestAction: PropTypes.func,
  singleRequestAction: PropTypes.func,
};

export const mapStateToProps = (state) => ({
  processRequest: state.processRequest,
});

export default connect(mapStateToProps, { processRequestAction, getManagerRequestAction, singleRequestAction })(ProcessRequest);
