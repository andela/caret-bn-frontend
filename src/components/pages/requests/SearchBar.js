import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Container, Row, Col, Form, Button,
} from 'react-bootstrap';
import { connect } from 'react-redux';
import DayPickerInput from 'react-day-picker/DayPickerInput';
import 'react-day-picker/lib/style.css';
import moment from 'moment';
import { getLocations } from '../../../actions/locationActions';
import { searchRequestAction } from '../../../actions/requestsActions';
import { managerSearchRequestAction } from '../../../actions/managerRequestAction';
import authHelper from '../../../helpers/authHelper';

const { checkManager } = authHelper;
export class SearchBar extends Component {
  state = {
    origin: '',
    destination: '',
    statusId: '',
    duration: '',
    departureDate: '',
    reasons: '',
    isLoading: false,
    openSearch: false,
    emptyParams: false,
  };

  componentDidMount() {
    const { getLocations } = this.props;
    getLocations();
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({
      [name]: value,
    });
  };

  handleDayChange = (departureDate) => {
    const dateValue = moment(departureDate).format('YYYY-MM-DD');
    this.setState((state) => ({ ...state, departureDate: dateValue }));
  }

  hideFilter = () => {
    const { state: { openSearch } } = this;
    this.setState((state) => ({ ...state, openSearch: !openSearch }));
  };

  searchRequest = async () => {
    const { state, props } = this;
    let searchParams = '';
    const {
      isLoading, openSearch, emptyParams, ...options
    } = state;

    Object.keys(options).forEach((key) => {
      if (options[`${key}`] !== '') {
        searchParams += `${key}=${options[`${key}`]}&`;
      }
    });

    searchParams = searchParams.substring(0, searchParams.length - 1);
    if (searchParams !== '') {
      searchParams = `?${searchParams}`;
      this.setState((state) => ({ ...state, isLoading: true }));
      if (!checkManager()) {
        await props.managerSearchRequestAction(searchParams);
      } else {
        await props.searchRequestAction(searchParams);
      }
      this.setState((state) => ({ ...state, isLoading: false, emptyParams: false }));
    } else {
      this.setState({ emptyParams: true });
    }
  }

  render() {
    const { props, state } = this;
    const {
      isLoading, openSearch, departureDate, emptyParams,
    } = state;
    const { locations } = props;
    return (
      <>
        <div className="search-box py-3">
          <Container>
            <Row>
              <Col xs={12} sm={6} md={4} lg={4}>
                <Button className="btn-block" style={{ width: '100%' }} onClick={() => this.hideFilter()} data-test="filter-request">
                  { !openSearch ? 'Search By...' : 'Hide Search'}
                </Button>
              </Col>
            </Row>

            { openSearch && (
              <>
                <Row className="search-bar">
                  { (!checkManager() && window.location.pathname !== '/requests') && (
                    <Col xs={12} sm={6} md={4} lg={4}>
                      <Form.Group>
                        <Form.Control type="text" name="username" onChange={this.handleChange} placeholder="Username..." minLength="10" maxLength="100" />
                      </Form.Group>
                    </Col>
                  )}
                  {(!checkManager() && window.location.pathname !== '/requests') ? (
                  <Col xs={12} sm={6} md={4} lg={4}>
                  <Form.Group>
                    <Form.Control as="select" name="origin" onChange={this.handleChange}>
                      <option selected disabled>Select Origin...</option>
                      <option value="">None</option>
                      {
                        locations ? locations.data.map((location) => <option key={location.id} value={location.name}>{location.name}</option>) : null
                      }
                    </Form.Control>
                  </Form.Group>
                  </Col>
                  ) : (
                    <Col xs={12} sm={6} md={6} lg={6}>
                  <Form.Group>
                    <Form.Control as="select" name="origin" onChange={this.handleChange}>
                      <option selected disabled>Select Origin...</option>
                      <option value="">None</option>
                      {
                        locations ? locations.data.map((location) => <option key={location.id} value={location.name}>{location.name}</option>) : null
                      }
                    </Form.Control>
                  </Form.Group>
                    </Col>
                  )}

                {(!checkManager() && window.location.pathname !== '/requests') ? (
                  <Col xs={12} sm={6} md={4} lg={4}>
                  <Form.Group>
                    <Form.Control as="select" name="destination" onChange={this.handleChange}>
                      <option selected disabled>Select Destination...</option>
                      <option value="">None</option>
                      {
                        locations ? locations.data.map((location) => <option key={location.id} value={location.name}>{location.name}</option>) : null
                      }
                    </Form.Control>
                  </Form.Group>
                  </Col>
                ) : (
                    <Col xs={12} sm={6} md={6} lg={6}>
                  <Form.Group>
                    <Form.Control as="select" name="destination" onChange={this.handleChange}>
                      <option selected disabled>Select Destination...</option>
                      <option value="">None</option>
                      {
                        locations ? locations.data.map((location) => <option key={location.id} value={location.name}>{location.name}</option>) : null
                      }
                    </Form.Control>
                  </Form.Group>
                    </Col>
                )}

                  <Col xs={12} sm={6} md={4} lg={4}>
                    <Form.Group>
                      <Form.Control as="select" onChange={this.handleChange} name="statusId" data-test="status-id">
                        <option selected disabled>Select Status...</option>
                        <option value="">None</option>
                        <option value={1}>1 - Pending</option>
                        <option value={2}>2 - Rejected</option>
                        <option value={3}>3 - Approved</option>
                      </Form.Control>
                    </Form.Group>
                  </Col>

                  <Col xs={12} sm={6} md={4} lg={4}>
                    <Form.Group>
                      <Form.Control type="number" onChange={this.handleChange} name="duration" min="1" max="999" data-test="duration" placeholder="Duration in days..." />
                    </Form.Group>
                  </Col>

                  <Col xs={12} sm={6} md={4} lg={4}>
                    <Form.Group className="day-picker-custom">
                      <DayPickerInput className="day-picker" inputProps={{ style: { border: 0, outline: 0 } }} placeholder="Departure Date..." value={departureDate} onDayChange={this.handleDayChange} />
                      {/* <i className="fa fa-times reset-cross" aria-hidden="true" onClick={this.resetDate} /> */}
                    </Form.Group>
                  </Col>

                  <Col xs={12} sm={6} md={8} lg={8}>
                    <Form.Group>
                      <Form.Control type="text" name="reasons" onChange={this.handleChange} placeholder="Reasons..." minLength="10" maxLength="100" />
                    </Form.Group>
                  </Col>

                  <Col xs={12} sm={6} md={4} lg={4}>
                    <Button className="btn-block" style={{ width: '100%' }} onClick={() => this.searchRequest()} data-test="search-request">
                      SEARCH
                    </Button>
                  </Col>
                  <Col xs={12} sm={6} md={4} lg={4}>
                    {emptyParams && (<small className="text-danger">Search options cannot be empty</small>)}
                  </Col>
                </Row>

                <Row>
                <Col xs={12} sm={6} md={4} lg={4}>
                  {!checkManager() ? (
                    <Button className="btn-block" style={{ width: '100%' }} onClick={() => window.location.replace('/user-manager')} data-test="all-requests-button">
                      All requests
                    </Button>
                  ) : (
                    <Button className="btn-block" style={{ width: '100%' }} onClick={() => window.location.replace('/requests')} data-test="all-requests-button">
                      All requests
                    </Button>
                  )}
                </Col>
                </Row>
              </>
            )}
          </Container>
        </div>

        <Container>
          <Row className="text-center mx-auto">
            {isLoading ? <i className="fas fa-spinner fa-pulse loader-big" /> : ''}
          </Row>
        </Container>
      </>
    );
  }
}

SearchBar.propTypes = {
  getLocations: PropTypes.func.isRequired,
  searchRequestAction: PropTypes.func.isRequired,
  managerSearchRequestAction: PropTypes.func.isRequired,
  locations: PropTypes.any,
};

export const mapStateToProps = (state) => ({
  locations: state.locations.data,
});

export default connect(mapStateToProps, { getLocations, searchRequestAction, managerSearchRequestAction })(SearchBar);
