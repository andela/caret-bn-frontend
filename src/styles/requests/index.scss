@import './destinations.scss';

.row {
	margin: 5px;
	padding: 5px;
}
.center-items {
	justify-content: space-around;
	padding: 0 10px;
}

.error-holder {
	display: flex;
	justify-content: center;
}

.row-between {
	justify-content: space-between;
	padding: 10px 25px;
}
.form-section {
	display: flex;
	justify-content: space-evenly;
	width: 100vw;
	flex-wrap: wrap;
	.select {
		width: 20vw;
		display: flex;
	}
}

.loading-componenent {
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
	height: 80vh;

	p {
		text-align: center;
		font-size: 12px;
		margin-top: 5px;
	}

	img {
		height: 80px;
	}
}

.section {
	display: flex;
	flex-direction: column;
	align-items: center;
	p {
		font-size: 15px;
		justify-content: center;
		text-align: center;
	}
}

@media (min-width: 768px) {
	.center-items {
		button {
			width: 150px;
			font-size: 11px;
		}
	}
}

@media (min-width: 1024px) {
	.center-items {
		button {
			width: 180px;
			font-size: 13px;
			font-weight: 400;
		}
	}
}

@media (min-width: 1024px) {
	.center-items {
		justify-content: space-around;
		padding: 0 50px;
	}
	.row {
		margin: 10px;
	}
	.space-horizontal {
		display: flex;
		justify-content: space-between;
	}
	.loading-componenent {
		p {
			font-size: 20px;
		}
	}
}
