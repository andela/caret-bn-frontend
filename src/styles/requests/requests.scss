.request-item {
  background-color: $white;
  .request-status {
    text-transform: uppercase;
  }
}
.loader-big {
  font-size: 100px;
  color: $primary;
  margin: 0 auto;
  margin-top: 5rem;
}
.search-box {
  background-color: $white;
  padding-top: 10px;
}
.day-picker-custom {
  display: block;
  width: 100%;
  height: calc(1.5em + 0.75rem + 2px);
  padding: 0.375rem 0.75rem;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #495057;
  background-color: #fff;
  background-clip: padding-box;
  border: 1px solid #ced4da;
  border-radius: 0.25rem;
  transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
}
.reset-cross:hover {
  cursor: pointer;
}
.process-request-button {
  text-transform: uppercase;
}
