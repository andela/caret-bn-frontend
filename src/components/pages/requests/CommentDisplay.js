/* eslint-disable react/button-has-type */
/* eslint-disable camelcase */
/* eslint-disable react/destructuring-assignment */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable eqeqeq */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { sortArrayDesd as sortComments } from 'tesla-error-handler';
import {
  Container, Row, Col, Button, Form, Spinner,
} from 'react-bootstrap';
import { connect } from 'react-redux';
import 'react-day-picker/lib/style.css';
import {
  setComment, getComment, deleteComment, editComment,
} from '../../../actions/commentAction';
import { singleRequestAction } from '../../../actions/requestsActions';
import trash from '../../../assets/images/trash.svg';
import edit from '../../../assets/images/edit.png';
import cancel from '../../../assets/images/cancel.jpg';
import isAuthenticated from '../../../helpers/isAuthenticated';
import 'react-confirm-alert/src/react-confirm-alert.css';
import Confirm from '../../global/Confirm';
import { CommentItem } from './CommentItem';

export class CommentDisplay extends Component {
  state = {
    isLoading: false,
    openComment: false,
    comment: '',
    isProcessing: false,
    editMode: false,
    commentId: null,
    draftComment: null,
    loadingButton: false,
    loadingButtonEdit: false,
    deleteLoading: false,
  };

  // eslint-disable-next-line react/sort-comp
  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({
      [name]: value,
    });
  };

  componentDidMount = async () => {
    this.retrieveComment();
  }

  retrieveComment = async () => {
    const { requestId } = this.props;
    await this.props.getComment(requestId);
  }

  hideComment = () => {
    const { state: { openComment } } = this;
    this.setState((state) => ({ ...state, openComment: !openComment }));
  };

  handleSubmit = async (e) => {
    const { props } = this;
    const { requestId, singleRequestAction } = this.props;
    e.preventDefault();
    const {
      comment,
    } = this.state;
    const userData = {
      comment,
    };
    this.setState({ comment: '', loadingButton: true });
    await props.setComment(userData, requestId);
    this.setState({ loadingButton: false });
    await props.singleRequestAction(requestId);
    await this.props.getComment(requestId);
  };

  render() {
    const {
      isLoading, openComment, comment, commentId, editMode, draftComment, loadingButton, loadingButtonEdit, deleteLoading,
    } = this.state;
    const {
      data, commentData, editData, profile, requestId,
    } = this.props;
    const userInfo = isAuthenticated();

    return (
      <>
        <div className="search-box py-3">
        <Container>
        <Col xs={12} sm={6} md={4} lg={4}>
          <Button data-test="comment-view" className="full-width-buttons" style={{ width: '100%' }} onClick={() => this.hideComment()}>
           Comments
          </Button>
        </Col>
        </Container>
      { openComment && (
        <div className="search-bar">
        <form className="mt-3 px-4" onSubmit={this.handleSubmit}>
        <div className="form-group row p-0 mb-0">
             <div className="col-sm-12">
             <div className="input-group shadow-sm mb-3 mt-3">
               <div className="input-group-prepend">
                 <span className="input-group-text bg-white p-1" id="basic-addon1">
                 {profile &&  <img src={this.props.profile.image || 'https://via.placeholder.com/450'} className="icon menu-photo" /> }
                 </span>
               </div>
                     <Form.Control as="textarea" rows="3" data-test="comment" type="text" className="form-control" onChange={this.handleChange} name="comment" placeholder="Enter a comment here" value={comment} required />
             </div>
             </div>
        </div>
        <Button type="submit">{loadingButton ? <i style={{ fontSize: '20px' }} className="fas fa-spinner fa-pulse" /> : 'Post'}</Button>
        </form>
          <Row />
          <div className="request-item mb-3 p-3">
          {data && sortComments(data && data.data).map((element) => (
            <CommentItem userInfo={userInfo} requestId={requestId} element={element} editComment={this.props.editComment} deleteComment={this.props.deleteComment} getComment={this.props.getComment} />
          ))}
          </div>
        </div>
      )}
        </div>
        <Container>
            {isLoading ? (
<div className="d-flex justify-content-center">
        <Spinner animation="grow" size="lg" variant="primary" />
</div>
            ) : ''}
        </Container>
      </>
    );
  }
}

CommentDisplay.propTypes = {
  setComment: PropTypes.func.isRequired,
  getComment: PropTypes.func.isRequired,
  deleteComment: PropTypes.func.isRequired,
  locations: PropTypes.any,
  editComment: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  dataError: state.requestComment.dataError,
  data: state.requestComment.data,
  commentError: state.requestComment.commentError,
  commentData: state.requestComment.commentData,
  deleteData: state.requestComment.deleteData,
  deleteError: state.requestComment.deleteError,
  editData: state.requestComment.editData,
  editError: state.requestComment.editError,
  status: state.requestComment.status,
  profile: state.profile.data.profile,
});

export default connect(mapStateToProps, {
  setComment, getComment, deleteComment, singleRequestAction, editComment,
})(CommentDisplay);
