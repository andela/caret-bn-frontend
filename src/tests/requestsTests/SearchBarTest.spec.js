import React from 'react';
import { shallow } from 'enzyme';
import { SearchBar, mapStateToProps } from '../../components/pages/requests/SearchBar';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../../reducers/index';
import requestsMocks from '../mocks/requestsMocks';
import DayPickerInput from 'react-day-picker/DayPickerInput';


const middlewares = [thunk];

const mainState = {
  locations: {
    dataError: null,
    data: [
      {
        id: 1,
        name: 'Kigali Office',
        country: 'Rwanda',
      },
      {
        id: 2,
        name: 'Lagos Office',
        country: 'Nigeria',
      },
    ],
  },
}

const props = {
  props: requestsMocks.getLocationsProps,
  getLocations: jest.fn(),
}

const testStore = (state) => {
  const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
  return createStoreWithMiddleware(rootReducer, state);
};

const setUp = (initialState = {}) => {
  const store = testStore(initialState);
  const wrapper = shallow(
    <SearchBar {...props} store={store} />
  );
  return wrapper;
}

describe('SearchBar Test Suite', () => {
  it('Should Mount Successfully', () => {
    const component = setUp(mainState);
    component.instance().hideStats();
    expect(component.exists()).toBe(true);
  });

  it('Should handleDayChange', () => {
    const component = setUp(mainState);
    component.find('.btn-block').first().simulate('click')
    component.find('.day-picker').first().simulate('dayChange')
    expect(component).toMatchSnapshot();

  });

  it('Should search for pending requests', () => {
    const component = setUp(mainState);
    const handleSubmitSpy = jest.spyOn(component.instance(), 'searchRequest');
    component.find('[data-test="filter-request"]').simulate('click');

    const statusId = { target: { name: 'statusId', value: '1' } };
    component.find('[data-test="status-id"]').simulate('change', statusId);
    component.find('[data-test="search-request"]').simulate('click');
    expect(handleSubmitSpy).toReturn();
  });

  it('Should return initial data', () => {
    const initialState = {
      locations: {
        dataError: null,
        data: null,
      },
    };
    expect(mapStateToProps(initialState).locations).toEqual(null);
  });
});
