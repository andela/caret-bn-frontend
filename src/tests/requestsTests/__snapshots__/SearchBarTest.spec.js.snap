// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Test Suite Should handleDayChange 1`] = `ShallowWrapper {}`;
